### ISSUE TITLE
Add XSPF support for playlist import and export

### ISSUE THREAD

**Issue by lazka** [quodlibet/quodlibet#1122]
Original [issue 1122](https://code.google.com/p/quodlibet/issues/detail?id=1122) created by shadowfax5124 on 2013-02-20T22:50:53.000Z:

While the m3u plugin suffices, I prefer the xspf. Are ther any plans to support it?

**Comment by declension** [comment]
There's also now JSPF: http://www.xspf.org/jspf/

**Comment by ibrokemypie** [comment]
No news for 2 years? Really would love metadata based playlists rather than file based...

**Comment by declension** [comment]
:heavy_check_mark:  I've got an XSPF-based playlist browser PoC working for QL 

Obviously, this is a deeper integration than just import and export support, but (and see discussion on #518), this could provide other benefits that now seem worth exploring.

### Things to think about
Even assuming it's something that fully works, and we'd want, then maybe have these first:
* [ ] Warning / documentation this is happening
* [ ] Ample thought into any knock-on effects in the remote "API"
* [ ] Automation around migration (see below)  

### Migrating users to XSPF-based playlists
1. Migrate all users to XSPF-based once, on upgrade of QL (either as a planned migration or just on first load of any non-XSPFs)
2. Read both formats, but save as XSPF only. Not many of the benefits (see #1010, #3105 etc) though until playlists _all_ have metadata (this is another open question)
3. Do 2., then 1 on the next (major?) release.
4. Have a preference system for "preferred playlist format"

### Unification of playlist support
 * It seems that m3u / pls should perhaps just be another supported format eventually that can be read natively if dropped into the directory (related: #174) 
* Then maybe remove plugins and expose UI to export or "add" in any format "natively"

**Comment by declension** [comment]
@lazka / @frestr any opinions?

**Comment by declension** [comment]
(See PoC in [`xspf-support`](..tree/xspf-support) branch)

**Comment by lazka** [comment]
What's the advantage of using XSPF to store things compared to the current way?

**Comment by declension** [comment]
> What's the advantage of using XSPF to store things compared to the current way?

#### TL; DR 
It's more about the things this can then unlock (#518, #1010, #3105 / #3131 / #708)

#### Longer version 
Not much, for now (other than this issue of course). 

But because this format allows us to store metadata _as well as_ the filename, we can have a chance of saving files have gone away / moved, e.g. intelligently "finding" missing playlist items by artist / title / album (currently) but presumably by say MBIDs or `labelid` in the future (it's extensible, or offers [`identifier`](http://www.xspf.org/xspf-v1.html#rfc.section.4.1.1.2.14.1.1.1.2) which perhaps we could use here)

I guess a few other things, e.g. comments / URLs / images for playlists would then be natively supported if extending the Playlist Browser... but not sure how useful these are. Though one thing I'd be quite interested in is a QL -> Spotify playlist bridge - see https://github.com/sepehr/xspfy/blob/master/xspfy.py actually.

**Comment by lazka** [comment]
OK, I was just wondering if I'm missing something.

re migration, the read many formats, write preferred when needed path (2) seems the safest to me.

**Comment by frestr** [comment]
I like the idea, and it does seem like it would make it eaiser to fix some of the playlist issues.

Regarding migration, I think migrating all playlists at once / on first read would be the simplest option. I would guess most users don't really care that much about how playlists are stored in the backend, and those who do (or say edit the files directly) would need to make updates anyway unless we're going to continue supporting the old format too.

It would also still be possible to export the playlist as m3u, if one wants a file in the old format (for some other purpose).

**Comment by derek-v-s** [comment]
@declension What's the status on this? I like the idea, and I'm interested in contributing.

**Comment by declension** [comment]
@derek-v-s It's basically working. I just never got round to merging, as it's quite a drastic change and got nervous about migrations

Please check out the branch though and try running it, this would be very helpful.

**Comment by derek-v-s** [comment]
No problems so far. 

- It converted my old playlist and put the old one in `.archive`. 
- A new playlist was created and loaded successfully. 
- A renamed file was removed from the playlist.

**Comment by derek-v-s** [comment]
Personally I'd like the xml to be human readable -- i.e. multiple lines with indentation.

https://stackoverflow.com/questions/3095434/inserting-newlines-in-xml-file-generated-via-xml-etree-elementtree-in-python

Thoughts?

**Comment by declension** [comment]
> Personally I'd like the xml to be human readable -- i.e. multiple lines with indentation.

Fair enough, though perhaps this is a nice-to-have that can be added once the feature is settled. Workaround: `xmllint`

Thanks for testing!

**Comment by derek-v-s** [comment]
Sounds fine. :-]

My pleasure.
